src/Sylius/Bundle/ResourceBundle/Controller/ResourceController.php

```diff
@@ -176,43 +236,49 @@ public function indexAction(Request $request)
      */
     public function createAction(Request $request)
     {
-        $this->isGrantedOr403('create');
+        $configuration = $this->configurationFactory->create($this->metadata, $request);
 
-        $resource = $this->createNew();
-        $form = $this->getForm($resource);
+        $this->isGrantedOr403($configuration, ResourceActions::CREATE);
 
-        if ($request->isMethod('POST') && $form->submit($request)->isValid()) {
-            $resource = $this->domainManager->create($form->getData());
+        $resource = $this->createNew($configuration);
+        $form = $this->createForm($configuration, $resource);
 
-            if ($this->config->isApiRequest()) {
-                if ($resource instanceof ResourceEvent) {
-                    throw new HttpException($resource->getErrorCode(), $resource->getMessage());
-                }
+        if ($request->isMethod('POST') && $form->submit($request)->isValid()) {
+            $this->eventDispatcher->dispatch(GenericResourceEvents::PRE_CREATE, new GenericResourceEvent($resource, $this->metadata, $configuration, ResourceActions::CREATE));
+            $event = $this->eventDispatcher->dispatchResourceEvent(ResourceEvents::PRE_CREATE, $resource);
 
-                return $this->handleView($this->view($resource, 201));
+            if ($event->isPropagationStopped() && !$configuration->isHtmlRequest()) {
+                throw new HttpException($event->getErrorCode(), $event->getMessage());
             }
 
-            if ($resource instanceof ResourceEvent) {
-                return $this->redirectHandler->redirectToIndex();
+            $this->manager->persist($resource);
```

Comment:
We loose here the manager that was dispatching events. It means that if I want to modify my resource in another action, I need to dispatch events myself + check if propagation has stopped? Is it done on purpose?

I guess a DispatcherAwareManager would be really convenient here.